### PR TITLE
RUMM-3185 Telemetry on Message Bus

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

The message-bus will support transmitting Telemetry messages. The core will proxy telemetry messages and will be able collect Configuration Telemetry while Features are lazily initialised.

After 5s, the core will automatically dispatch a Configuration Telemetry message so it can be reported in RUM.

### How?

The Core now implements a `MessageBus` class that will forward any messages to registered receivers, except if the message is a Configuration Telemetry. In that case, the configuration will be save so it can be reported later on.

The `RUMTelemetry` has been replaced by a `TelemetryReceiver` complying to `FeatureMessageReceiver`.

The `Telemetry` interface in the Internal module has been modified to support partial configuration report as follow:

```swift
let telemetry = TelemetryCore(core: core)

telemetry.configuration(
    batchSize: batchSize,
    batchUploadFrequency:batchUploadFrequency,
    useLocalEncryption: useLocalEncryption,
    useProxy: useProxy
)

telemetry.configuration(trackErrors: true)

telemetry.configuration(useTracing: true)
```

> The implies the definition of a `ConfigurationTelemetry` object in the Internal module.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

File: Sources/Datadog/DatadogCore/DatadogCore.swift [hunk 1]
```diff
@@ -112,6 +106,10 @@ internal final class DatadogCore {
         self.contextProvider.subscribe(\.version, to: applicationVersionPublisher)
         self.contextProvider.subscribe(\.trackingConsent, to: consentPublisher)
 
+        // connect the core to the message bus.
+        // the bus will keep a weak ref to the core.
+        bus.connect(core: self)
```
Comment (ncreated) [inline Sources/Datadog/DatadogCore/DatadogCore.swift:111]: non-blocking/ It's is quite critical piece of logic we're changing here - modifying the way in which the heart 💜 of `DatadogCore` is connected to the neural systems of Features 🙃 (message bus). With the lack of any [`L`-size unit test](https://github.com/DataDog/dd-sdk-ios/wiki/Test-pyramid-in-Datadog-iOS-SDK-project) defined in [DatadogCoreTests](https://github.com/DataDog/dd-sdk-ios/blob/feature/v2/Tests/DatadogTests/Datadog/DatadogCore/DatadogCoreTests.swift) it looks like an open surgery. No test is guarding for zero regression in this case (we're changing message-bus related tests along with this new implementation, so they don't count).

This is non-blocking ask, but I'm raising a flag of having very limited unit tests coverage for `DatadogCore`. I'd be up for pairing on adding some that cover message-bus communication 🙂.

**Review by ncreated — CHANGES_REQUESTED** [review]
It looks good and `TelemetryCore(core:)` is quite pleasant API from the Feature's POV 👌.

I left few comments, with one or two blocking due to lack of tests coverage for some of new APIs we're adding here.

I'd also love to see more tests being added for `DatadogCore` as we move forward with migration - that one is not blocking, but I volunteer for a pair programming on this some day 🙋.

File: Sources/Datadog/Datadog.swift [hunk 2]
```diff
@@ -253,8 +265,7 @@ public class Datadog {
             verbosityLevel: { Datadog.verbosityLevel }
         )
 
-        telemetry?.configuration(configuration: configuration)
-        DD.telemetry = telemetry ?? NOPTelemetry()
+        DD.telemetry = telemetry
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/Datadog.swift:268]: Indeed, I would like to remove it but it's quit an effort to remove. I've created [RUMM-3218](https://datadoghq.atlassian.net/browse/RUMM-3218) to tackle that.

[the remark this answers is not in the document]
Comment (ncreated) [reply Sources/Datadog/Datadog.swift:268]: Nice 👍. Yes, it's gonna be tedious task for existing Features. Not that scary for Session Replay, as we don't (yet) use telemetry there - the `TelemetryCore` abstraction is nice 👍.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift [hunk 3]
```diff
@@ -91,10 +140,118 @@ extension Telemetry {
...
     public init() { }
-    public func debug(id: String, message: String) {}
-    public func error(id: String, message: String, kind: String?, stack: String?) {}
+    /// no-op
+    public func send(telemetry: TelemetryMessage) { }
+}
+
+public struct TelemetryCore: Telemetry {
+    /// A weak core reference.
+    private weak var core: DatadogCoreProtocol?
```
[the remark this answers is not in the document]
Comment (maxep) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift:208]: Thanks for highlighting this, @ncreated! I've made some research and found out that `weak` ref is actually atomic: https://github.com/apple/swift/pull/1454

The issue you are facing is that the `core` defined in test is itself not thread-safe, not the `weak` reference in `TelemetryCore`. To properly test the weak reference deallocation atomicity, you can do:
```swift
@ReadWriteLock
var core: DatadogCoreProtocol?

func testItDoesNotReatinCoreReference() {
    for _ in (0..<10_000) {
        autoreleasepool {
            core = PassthroughCoreMock()
            let telemetry = TelemetryCore(core: core!)

            callConcurrently(closures: [
                { telemetry.debug("") },
                { telemetry.error("") },
                { self.core = nil },
            ], iterations: 2)
        }
    }
}
```

So the problem is not with the `weak` reference, but on how we retain the `core`. We would need to make sure that the upcoming core registry keeps atomic references to core instances, which is currently not the case with our `defaultDatadogCore` static var.

File: Sources/Datadog/DatadogCore/DatadogCore.swift (hunk 1 above)
Comment (maxep) [reply Sources/Datadog/DatadogCore/DatadogCore.swift:111]: I'm taking note on that, you are right 👍 We can brainstorm and pair on how to provide a good pattern for defining L-size units. I think the Integration Units target could be a good host for that. I will book a session!

File: DatadogInternal/Sources/Telemetry/Telemetry.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ncreated) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift:208]: Great deep dive @maxep 🎓👌 and it all makes senses 👍. The way `RUMM-3217` is phrased is to _"Ensure that SDK instance can be destroyed safely"_ and I believe it is still valid given what we conclude:

> We would need to make sure that the upcoming core registry keeps atomic references to core instances

**Review by ncreated — APPROVED** [review]
Well done & nice work 💪👌

**Pull request merged**